[macOS] CI build segfaults

This is an old problem, which I never managed to solve. Often times, macOS builds segfault for no apparent reason. 

```
2025-10-24T19:47:36.2288240Z   
2025-10-24T19:47:36.2288340Z     warn(stderr)
2025-10-24T19:47:36.2288420Z 
2025-10-24T19:47:36.2288730Z -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
2025-10-24T19:47:36.2289060Z ================ 376 passed, 318 skipped, 5 warnings in 24.69s =================
2025-10-24T19:47:36.3336590Z /Users/runner/work/_temp/dd665b0d-2f1e-4bbe-9da3-789bc1952906.sh: line 1:  4599 Segmentation fault: 11  "/Users/runner/work/_temp/cibw/bin/cibuildwheel" "." --output-dir "wheelhouse" 2>&1
2025-10-24T19:47:36.3350410Z ##[error]Process completed with exit code 139.
2025-10-24T19:47:36.3460480Z Post job cleanup.
2025-10-24T19:47:36.3525150Z Post job cleanup.
2025-10-24T19:47:36.8476810Z [command]/opt/homebrew/bin/git version
2025-10-24T19:47:36.8723010Z git version 2.50.1
2025-10-24T19:47:36.8740870Z Copying '/Users/runner/.gitconfig' to '/Users/runner/work/_temp/ff7a4887-d387-40bd-819f-b10a715a5894/.gitconfig'
2025-10-24T19:47:36.8747860Z Temporarily overriding HOME='/Users/runner/work/_temp/ff7a4887-d387-40bd-819f-b10a715a5894' before making global git config changes
2025-10-24T19:47:36.8748580Z Adding repository directory to the temporary git global config as a safe directory
2025-10-24T19:47:36.8750580Z [command]/opt/homebrew/bin/git config --global --add safe.directory /Users/runner/work/psutil/psutil
2025-10-24T19:47:36.8846710Z [command]/opt/homebrew/bin/git config --local --name-only --get-regexp core\.sshCommand
2025-10-24T19:47:36.9160700Z [command]/opt/homebrew/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'core\.sshCommand' && git config --local --unset-all 'core.sshCommand' || :"
2025-10-24T19:47:37.1356390Z [command]/opt/homebrew/bin/git config --local --name-only --get-regexp http\.https\:\/\/github\.com\/\.extraheader
2025-10-24T19:47:37.1404050Z http.https://github.com/.extraheader
2025-10-24T19:47:37.1409410Z [command]/opt/homebrew/bin/git config --local --unset-all http.https://github.com/.extraheader
2025-10-24T19:47:37.1635030Z [command]/opt/homebrew/bin/git submodule foreach --recursive sh -c "git config --local --name-only --get-regexp 'http\.https\:\/\/github\.com\/\.extraheader' && git config --local --unset-all 'http.https://github.com/.extraheader' || :"
2025-10-24T19:47:37.3066270Z Cleaning up orphan processes
```

--- a/psutil/arch/all/init.c
+++ b/psutil/arch/all/init.c
@@ -115,6 +115,16 @@ psutil_check_pid_range(PyObject *self, PyObject *args) {
 }
 
 
+// Use it when invalid args are passed to a C function.
+int
+psutil_badargs(const char *funcname) {
+    PyErr_Format(
+        PyExc_RuntimeError, "%s() invalid args passed to function", funcname
+    );
+    return -1;
+}
+
+
 // Called on module import on all platforms.
 int
 psutil_setup(void) {

--- a/psutil/arch/all/init.h
+++ b/psutil/arch/all/init.h
@@ -118,10 +118,11 @@ PyObject *psutil_PyErr_SetFromOSErrnoWithSyscall(const char *syscall);
     #endif
 #endif
 
+int psutil_badargs(const char *funcname);
+int psutil_setup(void);
+
 #if defined(PSUTIL_WINDOWS) || defined(PSUTIL_BSD) || defined(PSUTIL_OSX)
 PyObject *psutil_pids(PyObject *self, PyObject *args);
 #endif
-
 PyObject *psutil_set_debug(PyObject *self, PyObject *args);
 PyObject *psutil_check_pid_range(PyObject *self, PyObject *args);
-int psutil_setup(void);

--- a/psutil/arch/freebsd/proc_socks.c
+++ b/psutil/arch/freebsd/proc_socks.c
@@ -227,7 +227,7 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
 
     for (i = 0; i < cnt; i++) {
         int lport, rport, state;
-        char lip[200], rip[200];
+        char lip[INET6_ADDRSTRLEN], rip[INET6_ADDRSTRLEN];
         char path[PATH_MAX];
         int inseq;
         py_tuple = NULL;

--- a/psutil/arch/freebsd/sys_socks.c
+++ b/psutil/arch/freebsd/sys_socks.c
@@ -179,7 +179,7 @@ psutil_gather_inet(
         if ((inp->inp_vflag & INP_IPV6) && (include_v6 == 0))
             continue;
 
-        char lip[200], rip[200];
+        char lip[INET6_ADDRSTRLEN], rip[INET6_ADDRSTRLEN];
 
         xf = psutil_get_file_from_sock(so->xso_so, psutil_xfiles, psutil_nxfiles);
         if (xf == NULL)

--- a/psutil/arch/osx/cpu.c
+++ b/psutil/arch/osx/cpu.c
@@ -73,8 +73,9 @@ psutil_cpu_times(PyObject *self, PyObject *args) {
 
     error = host_statistics(mport, HOST_CPU_LOAD_INFO,
                             (host_info_t)&r_load, &count);
+    mach_port_deallocate(mach_task_self(), mport);
+
     if (error != KERN_SUCCESS) {
-        mach_port_deallocate(mach_task_self(), mport);
         return PyErr_Format(
             PyExc_RuntimeError,
             "host_statistics(HOST_CPU_LOAD_INFO) syscall failed: %s",
@@ -82,7 +83,6 @@ psutil_cpu_times(PyObject *self, PyObject *args) {
         );
     }
 
-    mach_port_deallocate(mach_task_self(), mport);
     return Py_BuildValue(
         "(dddd)",
         (double)r_load.cpu_ticks[CPU_STATE_USER] / CLK_TCK,
@@ -107,8 +107,9 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
     }
 
     ret = host_statistics(mport, HOST_VM_INFO, (host_info_t)&vmstat, &count);
+    mach_port_deallocate(mach_task_self(), mport);
+
     if (ret != KERN_SUCCESS) {
-        mach_port_deallocate(mach_task_self(), mport);
         PyErr_Format(
             PyExc_RuntimeError,
             "host_statistics(HOST_VM_INFO) failed: %s",
@@ -117,7 +118,6 @@ psutil_cpu_stats(PyObject *self, PyObject *args) {
         return NULL;
     }
 
-    mach_port_deallocate(mach_task_self(), mport);
     return Py_BuildValue(
         "IIIII",
         vmstat.v_swtch,
@@ -197,10 +197,11 @@ psutil_cpu_freq(PyObject *self, PyObject *args) {
     uint32_t pMin = 0, eMin = 0, min = 0, max = 0, curr = 0;
 
     if (!psutil_find_pmgr_entry(&entry)) {
-        return PyErr_Format(
+        PyErr_SetString(
             PyExc_RuntimeError,
             "'pmgr' entry not found in AppleARMIODevice"
         );
+        return NULL;
     }
 
     pCoreRef = IORegistryEntryCreateCFProperty(
@@ -236,9 +237,6 @@ cleanup:
         CFRelease(eCoreRef);
     if (entry != IO_OBJECT_NULL)
         IOObjectRelease(entry);
-
-    if (PyErr_Occurred())
-        return NULL;
 
     return Py_BuildValue(
         "KKK",
@@ -302,11 +300,17 @@ psutil_per_cpu_times(PyObject *self, PyObject *args) {
         goto error;
     }
 
-    error = host_processor_info(mport, PROCESSOR_CPU_LOAD_INFO, &cpu_count, &info_array, &info_count);
+    error = host_processor_info(
+        mport, PROCESSOR_CPU_LOAD_INFO, &cpu_count, &info_array, &info_count
+    );
     mach_port_deallocate(mach_task_self(), mport);
 
     if (error != KERN_SUCCESS || !info_array) {
-        PyErr_Format(PyExc_RuntimeError, "host_processor_info failed: %s", mach_error_string(error));
+        PyErr_Format(
+            PyExc_RuntimeError,
+            "host_processor_info failed: %s",
+            mach_error_string(error)
+        );
         goto error;
     }
 
@@ -338,10 +342,11 @@ psutil_per_cpu_times(PyObject *self, PyObject *args) {
 
 error:
     Py_XDECREF(py_retlist);
-    vm_deallocate(
-        mach_task_self(),
-        (vm_address_t)info_array,
-        info_count * sizeof(integer_t)
-    );
+    if (info_array)
+        vm_deallocate(
+            mach_task_self(),
+            (vm_address_t)info_array,
+            info_count * sizeof(integer_t)
+        );
     return NULL;
 }

--- a/psutil/arch/osx/cpu.c
+++ b/psutil/arch/osx/cpu.c
@@ -329,6 +329,7 @@ psutil_per_cpu_times(PyObject *self, PyObject *args) {
         }
         if (PyList_Append(py_retlist, py_cputime)) {
             Py_DECREF(py_cputime);
+            goto error;
         }
         Py_DECREF(py_cputime);
     }

--- a/psutil/arch/osx/disk.c
+++ b/psutil/arch/osx/disk.c
@@ -188,6 +188,8 @@ psutil_disk_usage_used(PyObject *self, PyObject *args) {
 
     attrs.bitmapcount = ATTR_BIT_MAP_COUNT;
     attrs.volattr = ATTR_VOL_INFO | ATTR_VOL_SPACEUSED;
+    attrbuf.size = sizeof(attrbuf);
+
     Py_BEGIN_ALLOW_THREADS
     ret = getattrlist(mount_point, &attrs, &attrbuf, sizeof(attrbuf), 0);
     Py_END_ALLOW_THREADS

--- a/psutil/arch/osx/mem.c
+++ b/psutil/arch/osx/mem.c
@@ -89,14 +89,14 @@ PyObject *
 psutil_swap_mem(PyObject *self, PyObject *args) {
     int mib[2];
     struct xsw_usage totals;
-    vm_statistics64_data_t  vmstat;
+    vm_statistics64_data_t vmstat;
     long pagesize = psutil_getpagesize();
 
     mib[0] = CTL_VM;
     mib[1] = VM_SWAPUSAGE;
 
     if (psutil_sysctl(mib, 2, &totals, sizeof(totals)) != 0)
-        return psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl(HW_CPU_FREQ)");
+        return psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl(VM_SWAPUSAGE)");
 
     if (psutil_sys_vminfo(&vmstat) != 0)
         return NULL;
@@ -107,5 +107,6 @@ psutil_swap_mem(PyObject *self, PyObject *args) {
         (unsigned long long) totals.xsu_used,
         (unsigned long long) totals.xsu_avail,
         (unsigned long long) vmstat.pageins * pagesize,
-        (unsigned long long) vmstat.pageouts * pagesize);
+        (unsigned long long) vmstat.pageouts * pagesize
+    );
 }

--- a/psutil/arch/osx/net.c
+++ b/psutil/arch/osx/net.c
@@ -43,9 +43,8 @@ psutil_net_io_counters(PyObject *self, PyObject *args) {
     lim = buf + len;
 
     for (next = buf; next < lim; ) {
-        // Check we have enough space for if_msghdr.
         if ((size_t)(lim - next) < sizeof(struct if_msghdr)) {
-            psutil_debug("struct xfile size mismatch (skip entry)");
+            psutil_debug("struct if_msghdr size mismatch (skip entry)");
             break;
         }
 
@@ -74,13 +73,12 @@ psutil_net_io_counters(PyObject *self, PyObject *args) {
                 continue;
             }
 
-            char ifc_name[32];
+            char ifc_name[IFNAMSIZ];
             size_t namelen = sdl->sdl_nlen;
+            if (namelen >= IFNAMSIZ)
+                namelen = IFNAMSIZ - 1;
 
-            if (namelen >= sizeof(ifc_name))
-                namelen = sizeof(ifc_name) - 1;
-
-            strncpy(ifc_name, sdl->sdl_data, namelen);
+            memcpy(ifc_name, sdl->sdl_data, namelen);
             ifc_name[namelen] = '\0';
 
             py_ifc_info = Py_BuildValue(
@@ -101,6 +99,7 @@ psutil_net_io_counters(PyObject *self, PyObject *args) {
                 Py_CLEAR(py_ifc_info);
                 goto error;
             }
+
             Py_CLEAR(py_ifc_info);
         }
     }

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -796,12 +796,11 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
         fdp_pointer = &fds_pointer[i];
 
         if (fdp_pointer->proc_fdtype == PROX_FDTYPE_SOCKET) {
-            errno = 0;
             nb = proc_pidfdinfo(pid, fdp_pointer->proc_fd,
                                 PROC_PIDFDSOCKETINFO, &si, sizeof(si));
 
             // --- errors checking
-            if ((nb <= 0) || (nb < sizeof(si)) || (errno != 0)) {
+            if ((nb <= 0) || (nb < sizeof(si))) {
                 if (errno == EBADF) {
                     // let's assume socket has been closed
                     psutil_debug("proc_pidfdinfo(PROC_PIDFDSOCKETINFO) -> "
@@ -914,7 +913,6 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
                 if (!py_raddr)
                     goto error;
 
-                // construct the python list
                 py_tuple = Py_BuildValue(
                     "(iiiNNi)", fd, family, type, py_laddr, py_raddr, state);
                 if (!py_tuple)
@@ -932,7 +930,7 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
                     si.psi.soi_proto.pri_un.unsi_caddr.ua_sun.sun_path);
                 if (!py_raddr)
                     goto error;
-                // construct the python list
+
                 py_tuple = Py_BuildValue(
                     "(iiiOOi)",
                     fd, family, type,

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -561,9 +561,7 @@ psutil_proc_memory_uss(PyObject *self, PyObject *args) {
 PyObject *
 psutil_proc_threads(PyObject *self, PyObject *args) {
     pid_t pid;
-    int err, ret;
     kern_return_t kr;
-    unsigned int info_count = TASK_BASIC_INFO_COUNT;
     mach_port_t task = MACH_PORT_NULL;
     struct task_basic_info tasks_info;
     thread_act_port_array_t thread_list = NULL;
@@ -583,12 +581,11 @@ psutil_proc_threads(PyObject *self, PyObject *args) {
     if (psutil_task_for_pid(pid, &task) != 0)
         goto error;
 
-    info_count = TASK_BASIC_INFO_COUNT;
-    err = task_info(task, TASK_BASIC_INFO, (task_info_t)&tasks_info,
-                    &info_count);
-    if (err != KERN_SUCCESS) {
-        // errcode 4 is "invalid argument" (access denied)
-        if (err == 4) {
+    // Get basic task info (optional, ignored if access denied)
+    mach_msg_type_number_t info_count = TASK_BASIC_INFO_COUNT;
+    kr = task_info(task, TASK_BASIC_INFO, (task_info_t)&tasks_info, &info_count);
+    if (kr != KERN_SUCCESS) {
+        if (kr == KERN_INVALID_ARGUMENT) {
             AccessDenied("task_info(TASK_BASIC_INFO)");
         }
         else {
@@ -599,8 +596,8 @@ psutil_proc_threads(PyObject *self, PyObject *args) {
         goto error;
     }
 
-    err = task_threads(task, &thread_list, &thread_count);
-    if (err != KERN_SUCCESS) {
+    kr = task_threads(task, &thread_list, &thread_count);
+    if (kr != KERN_SUCCESS) {
         PyErr_Format(PyExc_RuntimeError, "task_threads() syscall failed");
         goto error;
     }
@@ -631,30 +628,28 @@ psutil_proc_threads(PyObject *self, PyObject *args) {
         Py_CLEAR(py_tuple);
     }
 
-    ret = vm_deallocate(
-        task, (vm_address_t)thread_list, thread_count * sizeof(thread_act_t)
-    );
-
-    if (ret != KERN_SUCCESS)
-        PyErr_WarnEx(PyExc_RuntimeWarning, "vm_deallocate() failed", 2);
-
-    mach_port_deallocate(mach_task_self(), task);
+    if (thread_list != NULL) {
+        vm_deallocate(mach_task_self(), (vm_address_t)thread_list,
+                      thread_count * sizeof(thread_act_t));
+    }
+    if (task != MACH_PORT_NULL) {
+        mach_port_deallocate(mach_task_self(), task);
+    }
 
     return py_retlist;
 
 error:
-    if (task != MACH_PORT_NULL)
-        mach_port_deallocate(mach_task_self(), task);
     Py_XDECREF(py_tuple);
-    Py_DECREF(py_retlist);
-    if (thread_list != NULL) {
-        ret = vm_deallocate(
-            task, (vm_address_t)thread_list, thread_count * sizeof(thread_act_t)
-        );
+    Py_XDECREF(py_retlist);
 
-        if (ret != KERN_SUCCESS)
-            PyErr_WarnEx(PyExc_RuntimeWarning, "vm_deallocate() failed", 2);
+    if (thread_list != NULL) {
+        vm_deallocate(mach_task_self(), (vm_address_t)thread_list,
+                      thread_count * sizeof(thread_act_t));
     }
+    if (task != MACH_PORT_NULL) {
+        mach_port_deallocate(mach_task_self(), task);
+    }
+
     return NULL;
 }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -73,6 +73,9 @@ static int
 is_zombie(size_t pid) {
     struct kinfo_proc kp;
 
+    if (pid < 0)
+        return psutil_badargs("is_zombie");
+
     if (psutil_get_kinfo_proc(pid, &kp) == -1) {
         PyErr_Clear();
         return 0;

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -106,7 +106,7 @@ psutil_sysctl_procargs(pid_t pid, char *procargs, size_t *argmax) {
         }
 
         if (errno == EINVAL) {
-            psutil_debug("sysctl(KERN_PROCARGS2) -> EINVAL translated to NSP");
+            psutil_debug("sysctl(KERN_PROCARGS2) -> EINVAL translated to AD");
             AccessDenied("sysctl(KERN_PROCARGS2) -> EINVAL");
             return -1;
         }
@@ -130,20 +130,25 @@ psutil_sysctl_procargs(pid_t pid, char *procargs, size_t *argmax) {
  */
 static int
 psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
-    errno = 0;
     int ret;
 
+    if (!pti || size <= 0) {
+        errno = EINVAL;
+        return 0;
+    }
+
+    errno = 0;
     ret = proc_pidinfo(pid, flavor, arg, pti, size);
     if (ret <= 0) {
         psutil_raise_for_pid(pid, "proc_pidinfo()");
         return 0;
     }
 
-    // use the actual buffer size
+    // check for truncated return size
     if (ret < size) {
         psutil_raise_for_pid(
             pid,
-            "proc_pidinfo() returned size smaller than expected buffer size"
+            "proc_pidinfo() returned less data than requested buffer size"
         );
         return 0;
     }

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -49,6 +49,13 @@ psutil_get_kinfo_proc(pid_t pid, struct kinfo_proc *kp) {
     mib[2] = KERN_PROC_PID;
     mib[3] = pid;
 
+    if (pid < 0 || !kp) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_get_kinfo_proc() invalid args"
+        );
+        return -1;
+    }
+
     len = sizeof(struct kinfo_proc);
 
     if (sysctl(mib, 4, kp, &len, NULL, 0) == -1) {
@@ -89,8 +96,10 @@ psutil_sysctl_procargs(pid_t pid, char *procargs, size_t *argmax) {
     mib[1] = KERN_PROCARGS2;
     mib[2] = pid;
 
-    if (!procargs || !argmax || *argmax == 0) {
-        errno = EINVAL;
+    if (pid < 0 || !procargs || !argmax || *argmax == 0) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_sysctl_procargs() invalid args"
+        );
         return -1;
     }
 
@@ -132,8 +141,10 @@ static int
 psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
     int ret;
 
-    if (!pti || size <= 0) {
-        errno = EINVAL;
+    if (pid < 0  || !pti || size <= 0) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_proc_pidinfo() invalid args"
+        );
         return 0;
     }
 
@@ -175,8 +186,10 @@ static int
 psutil_task_for_pid(pid_t pid, mach_port_t *task) {
     kern_return_t err;
 
-    if (!task) {
-        errno = EINVAL;
+    if (pid < 0 || !task) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_task_for_pid() invalid args"
+        );
         return -1;
     }
 
@@ -214,8 +227,10 @@ psutil_proc_list_fds(pid_t pid, int *num_fds) {
     int max_size = 24 * 1024 * 1024;  // 24M
     struct proc_fdinfo *fds_pointer = NULL;
 
-    if (num_fds == NULL) {
-        errno = EINVAL;
+    if (pid < 0 || num_fds == NULL) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_proc_list_fds() invalid args"
+        );
         return NULL;
     }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -287,8 +287,8 @@ PyObject *
 psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
     pid_t pid;
     struct kinfo_proc kp;
-    PyObject *py_name;
-    PyObject *py_retlist;
+    PyObject *py_name = NULL;
+    PyObject *py_retlist = NULL;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
         return NULL;
@@ -300,6 +300,7 @@ psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
         // Likely a decoding error. We don't want to fail the whole
         // operation. The python module may retry with proc_name().
         PyErr_Clear();
+        Py_INCREF(Py_None);  // safely hold a reference
         py_name = Py_None;
     }
 
@@ -318,10 +319,8 @@ psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
         py_name                                    // (pystr) name
     );
 
-    if (py_retlist != NULL) {
-        // XXX shall we decref() also in case of Py_BuildValue() error?
-        Py_DECREF(py_name);
-    }
+    Py_DECREF(py_name);  // safe now, even if py_name was None
+
     return py_retlist;
 }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -825,8 +825,7 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
 
             //
             int fd, family, type, lport, rport, state;
-            // TODO: use INET6_ADDRSTRLEN instead of 200
-            char lip[200], rip[200];
+            char lip[INET6_ADDRSTRLEN], rip[INET6_ADDRSTRLEN];
             int inseq;
             PyObject *py_family;
             PyObject *py_type;

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -49,10 +49,8 @@ psutil_get_kinfo_proc(pid_t pid, struct kinfo_proc *kp) {
     mib[2] = KERN_PROC_PID;
     mib[3] = pid;
 
-    // fetch the info with sysctl()
     len = sizeof(struct kinfo_proc);
 
-    // now read the data from sysctl
     if (sysctl(mib, 4, kp, &len, NULL, 0) == -1) {
         // raise an exception and throw errno as the error
         psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl");

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -49,12 +49,8 @@ psutil_get_kinfo_proc(pid_t pid, struct kinfo_proc *kp) {
     mib[2] = KERN_PROC_PID;
     mib[3] = pid;
 
-    if (pid < 0 || !kp) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_get_kinfo_proc() invalid args"
-        );
-        return -1;
-    }
+    if (pid < 0 || !kp)
+        return psutil_badargs("psutil_get_kinfo_proc");
 
     len = sizeof(struct kinfo_proc);
 
@@ -96,12 +92,8 @@ psutil_sysctl_procargs(pid_t pid, char *procargs, size_t *argmax) {
     mib[1] = KERN_PROCARGS2;
     mib[2] = pid;
 
-    if (pid < 0 || !procargs || !argmax || *argmax == 0) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_sysctl_procargs() invalid args"
-        );
-        return -1;
-    }
+    if (pid < 0 || !procargs || !argmax || *argmax == 0)
+        return psutil_badargs("psutil_sysctl_procargs");
 
     if (sysctl(mib, 3, procargs, argmax, NULL, 0) < 0) {
         if (psutil_pid_exists(pid) == 0) {
@@ -141,18 +133,14 @@ static int
 psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
     int ret;
 
-    if (pid < 0  || !pti || size <= 0) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_proc_pidinfo() invalid args"
-        );
-        return 0;
-    }
+    if (pid < 0 || !pti || size <= 0)
+        return psutil_badargs("psutil_proc_pidinfo");
 
     errno = 0;
     ret = proc_pidinfo(pid, flavor, arg, pti, size);
     if (ret <= 0) {
         psutil_raise_for_pid(pid, "proc_pidinfo()");
-        return 0;
+        return -1;
     }
 
     // check for truncated return size
@@ -161,10 +149,10 @@ psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
             pid,
             "proc_pidinfo() returned less data than requested buffer size"
         );
-        return 0;
+        return -1;
     }
 
-    return ret;
+    return 0;
 }
 
 
@@ -186,12 +174,8 @@ static int
 psutil_task_for_pid(pid_t pid, mach_port_t *task) {
     kern_return_t err;
 
-    if (pid < 0 || !task) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_task_for_pid() invalid args"
-        );
-        return -1;
-    }
+    if (pid < 0 || !task)
+        return psutil_badargs("psutil_task_for_pid");
 
     err = task_for_pid(mach_task_self(), pid, task);
     if (err != KERN_SUCCESS) {
@@ -228,9 +212,7 @@ psutil_proc_list_fds(pid_t pid, int *num_fds) {
     struct proc_fdinfo *fds_pointer = NULL;
 
     if (pid < 0 || num_fds == NULL) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_proc_list_fds() invalid args"
-        );
+        psutil_badargs("psutil_proc_list_fds");
         return NULL;
     }
 
@@ -374,7 +356,7 @@ psutil_proc_pidtaskinfo_oneshot(PyObject *self, PyObject *args) {
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
         return NULL;
-    if (psutil_proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)) <= 0)
+    if (psutil_proc_pidinfo(pid, PROC_PIDTASKINFO, 0, &pti, sizeof(pti)) != 0)
         return NULL;
 
     total_user = pti.pti_total_user * PSUTIL_MACH_TIMEBASE_INFO.numer;
@@ -433,7 +415,7 @@ psutil_proc_cwd(PyObject *self, PyObject *args) {
         return NULL;
 
     if (psutil_proc_pidinfo(
-            pid, PROC_PIDVNODEPATHINFO, 0, &pathinfo, sizeof(pathinfo)) <= 0)
+            pid, PROC_PIDVNODEPATHINFO, 0, &pathinfo, sizeof(pathinfo)) != 0)
     {
         return NULL;
     }

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -760,6 +760,7 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
     struct proc_fdinfo *fds_pointer = NULL;
     struct proc_fdinfo *fdp_pointer;
     struct socket_fdinfo si;
+    const char* ntopret;
     PyObject *py_retlist = PyList_New(0);
     PyObject *py_tuple = NULL;
     PyObject *py_laddr = NULL;
@@ -848,32 +849,48 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
 
             if ((family == AF_INET) || (family == AF_INET6)) {
                 if (family == AF_INET) {
-                    inet_ntop(AF_INET,
-                              &si.psi.soi_proto.pri_tcp.tcpsi_ini. \
-                                  insi_laddr.ina_46.i46a_addr4,
-                              lip,
-                              sizeof(lip));
-                    inet_ntop(AF_INET,
-                              &si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_faddr. \
-                                  ina_46.i46a_addr4,
-                              rip,
-                              sizeof(rip));
+                    ntopret = inet_ntop(
+                        AF_INET,
+                        &si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_laddr.ina_46.i46a_addr4,
+                        lip,
+                        sizeof(lip)
+                    );
+                    if (!ntopret) {
+                        psutil_PyErr_SetFromOSErrnoWithSyscall("inet_ntop()");
+                        goto error;
+                    }
+                    ntopret = inet_ntop(
+                        AF_INET,
+                        &si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_faddr.ina_46.i46a_addr4,
+                        rip,
+                        sizeof(rip)
+                    );
+                    if (!ntopret) {
+                        psutil_PyErr_SetFromOSErrnoWithSyscall("inet_ntop()");
+                        goto error;
+                    }
                 }
                 else {
-                    inet_ntop(AF_INET6,
-                              &si.psi.soi_proto.pri_tcp.tcpsi_ini. \
-                                  insi_laddr.ina_6,
-                              lip, sizeof(lip));
-                    inet_ntop(AF_INET6,
-                              &si.psi.soi_proto.pri_tcp.tcpsi_ini. \
-                                  insi_faddr.ina_6,
-                              rip, sizeof(rip));
-                }
-
-                // check for inet_ntop failures
-                if (errno != 0) {
-                    psutil_PyErr_SetFromOSErrnoWithSyscall("inet_ntop()");
-                    goto error;
+                    ntopret = inet_ntop(
+                        AF_INET6,
+                        &si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_laddr.ina_6,
+                        lip,
+                        sizeof(lip)
+                    );
+                    if (!ntopret) {
+                        psutil_PyErr_SetFromOSErrnoWithSyscall("inet_ntop()");
+                        goto error;
+                    }
+                    ntopret = inet_ntop(
+                        AF_INET6,
+                        &si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_faddr.ina_6,
+                        rip,
+                        sizeof(rip)
+                    );
+                    if (!ntopret) {
+                        psutil_PyErr_SetFromOSErrnoWithSyscall("inet_ntop()");
+                        goto error;
+                    }
                 }
 
                 lport = ntohs(si.psi.soi_proto.pri_tcp.tcpsi_ini.insi_lport);

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -403,9 +403,11 @@ psutil_proc_name(PyObject *self, PyObject *args) {
         return NULL;
     if (psutil_get_kinfo_proc(pid, &kp) == -1)
         return NULL;
-    return PyUnicode_DecodeFSDefault(kp.kp_proc.p_comm);
-}
 
+    return PyUnicode_DecodeFSDefaultAndSize(
+        kp.kp_proc.p_comm, sizeof(kp.kp_proc.p_comm)
+    );
+}
 
 /*
  * Return process current working directory.

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -427,7 +427,9 @@ psutil_proc_cwd(PyObject *self, PyObject *args) {
         return NULL;
     }
 
-    return PyUnicode_DecodeFSDefault(pathinfo.pvi_cdir.vip_path);
+    return PyUnicode_DecodeFSDefaultAndSize(
+        pathinfo.pvi_cdir.vip_path, sizeof(pathinfo.pvi_cdir.vip_path)
+    );
 }
 
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -463,7 +463,7 @@ psutil_proc_exe(PyObject *self, PyObject *args) {
             return NULL;
         }
     }
-    return PyUnicode_DecodeFSDefault(buf);
+    return PyUnicode_DecodeFSDefaultAndSize(buf, ret);
 }
 
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -89,6 +89,11 @@ psutil_sysctl_procargs(pid_t pid, char *procargs, size_t *argmax) {
     mib[1] = KERN_PROCARGS2;
     mib[2] = pid;
 
+    if (!procargs || !argmax || *argmax == 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
     if (sysctl(mib, 3, procargs, argmax, NULL, 0) < 0) {
         if (psutil_pid_exists(pid) == 0) {
             NoSuchProcess("psutil_pid_exists -> 0");

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -135,11 +135,16 @@ psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
         psutil_raise_for_pid(pid, "proc_pidinfo()");
         return 0;
     }
-    if ((unsigned long)ret < sizeof(pti)) {
+
+    // use the actual buffer size
+    if (ret < size) {
         psutil_raise_for_pid(
-            pid, "proc_pidinfo() return size < sizeof(struct_pointer)");
+            pid,
+            "proc_pidinfo() returned size smaller than expected buffer size"
+        );
         return 0;
     }
+
     return ret;
 }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -839,11 +839,16 @@ psutil_proc_net_connections(PyObject *self, PyObject *args) {
             py_family = PyLong_FromLong((long)family);
             inseq = PySequence_Contains(py_af_filter, py_family);
             Py_DECREF(py_family);
+            if (inseq == -1)
+                goto error;
             if (inseq == 0)
                 continue;
+
             py_type = PyLong_FromLong((long)type);
             inseq = PySequence_Contains(py_type_filter, py_type);
             Py_DECREF(py_type);
+            if (inseq == -1)
+                goto error;
             if (inseq == 0)
                 continue;
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -1037,7 +1037,7 @@ psutil_proc_cmdline(PyObject *self, PyObject *args) {
 
     // read argmax and allocate memory for argument space.
     argmax = psutil_sysctl_argmax();
-    if (! argmax)
+    if (argmax == 0)
         goto error;
 
     procargs = (char *)malloc(argmax);

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -165,26 +165,30 @@ psutil_proc_pidinfo(pid_t pid, int flavor, uint64_t arg, void *pti, int size) {
  * - for PIDs != getpid() or PIDs which are not members of the procmod
  *   it requires root
  * As such we can only guess what the heck went wrong and fail either
- * with NoSuchProcess or giveup with AccessDenied.
- * Here's some history:
+ * with NoSuchProcess or give up with AccessDenied.
+ * References:
  * https://github.com/giampaolo/psutil/issues/1181
  * https://github.com/giampaolo/psutil/issues/1209
  * https://github.com/giampaolo/psutil/issues/1291#issuecomment-396062519
  */
 static int
 psutil_task_for_pid(pid_t pid, mach_port_t *task) {
-    // See: https://github.com/giampaolo/psutil/issues/1181
-    kern_return_t err = KERN_SUCCESS;
+    kern_return_t err;
+
+    if (!task) {
+        errno = EINVAL;
+        return -1;
+    }
 
     err = task_for_pid(mach_task_self(), pid, task);
     if (err != KERN_SUCCESS) {
         if (psutil_pid_exists(pid) == 0) {
             NoSuchProcess("task_for_pid");
         }
-        // Now done in Python.
-        // else if (psutil_is_zombie(pid) == 1)
-        //     PyErr_SetString(ZombieProcessError,
-        //                     "task_for_pid -> psutil_is_zombie -> 1");
+        else if (is_zombie(pid) == 1) {
+            PyErr_SetString(ZombieProcessError,
+                            "task_for_pid -> psutil_is_zombie -> 1");
+        }
         else {
             psutil_debug(
                 "task_for_pid() failed (pid=%ld, err=%i, errno=%i, msg='%s'); "
@@ -194,6 +198,7 @@ psutil_task_for_pid(pid_t pid, mach_port_t *task) {
         }
         return -1;
     }
+
     return 0;
 }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -313,9 +313,7 @@ psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
     if (psutil_get_kinfo_proc(pid, &kp) == -1)
         return NULL;
 
-    py_name = PyUnicode_DecodeFSDefaultAndSize(
-        kp.kp_proc.p_comm, sizeof(kp.kp_proc.p_comm)
-    );
+    py_name = PyUnicode_DecodeFSDefault(kp.kp_proc.p_comm);
     if (! py_name) {
         // Likely a decoding error. We don't want to fail the whole
         // operation. The python module may retry with proc_name().
@@ -404,9 +402,7 @@ psutil_proc_name(PyObject *self, PyObject *args) {
     if (psutil_get_kinfo_proc(pid, &kp) == -1)
         return NULL;
 
-    return PyUnicode_DecodeFSDefaultAndSize(
-        kp.kp_proc.p_comm, sizeof(kp.kp_proc.p_comm)
-    );
+    return PyUnicode_DecodeFSDefault(kp.kp_proc.p_comm);
 }
 
 /*
@@ -427,9 +423,7 @@ psutil_proc_cwd(PyObject *self, PyObject *args) {
         return NULL;
     }
 
-    return PyUnicode_DecodeFSDefaultAndSize(
-        pathinfo.pvi_cdir.vip_path, sizeof(pathinfo.pvi_cdir.vip_path)
-    );
+    return PyUnicode_DecodeFSDefault(pathinfo.pvi_cdir.vip_path);
 }
 
 
@@ -463,7 +457,7 @@ psutil_proc_exe(PyObject *self, PyObject *args) {
             return NULL;
         }
     }
-    return PyUnicode_DecodeFSDefaultAndSize(buf, ret);
+    return PyUnicode_DecodeFSDefaultAndSize(buf);
 }
 
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -313,17 +313,19 @@ psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
     if (psutil_get_kinfo_proc(pid, &kp) == -1)
         return NULL;
 
-    py_name = PyUnicode_DecodeFSDefault(kp.kp_proc.p_comm);
+    py_name = PyUnicode_DecodeFSDefaultAndSize(
+        kp.kp_proc.p_comm, sizeof(kp.kp_proc.p_comm)
+    );
     if (! py_name) {
         // Likely a decoding error. We don't want to fail the whole
         // operation. The python module may retry with proc_name().
         PyErr_Clear();
-        Py_INCREF(Py_None);  // safely hold a reference
+        Py_INCREF(Py_None);
         py_name = Py_None;
     }
 
     py_retlist = Py_BuildValue(
-        _Py_PARSE_PID "llllllidiO",
+        _Py_PARSE_PID "llllllldiO",
         kp.kp_eproc.e_ppid,                        // (pid_t) ppid
         (long)kp.kp_eproc.e_pcred.p_ruid,          // (long) real uid
         (long)kp.kp_eproc.e_ucred.cr_uid,          // (long) effective uid
@@ -331,14 +333,13 @@ psutil_proc_kinfo_oneshot(PyObject *self, PyObject *args) {
         (long)kp.kp_eproc.e_pcred.p_rgid,          // (long) real gid
         (long)kp.kp_eproc.e_ucred.cr_groups[0],    // (long) effective gid
         (long)kp.kp_eproc.e_pcred.p_svgid,         // (long) saved gid
-        kp.kp_eproc.e_tdev,                        // (int) tty nr
+        (long long)kp.kp_eproc.e_tdev,             // (long long) tty nr
         PSUTIL_TV2DOUBLE(kp.kp_proc.p_starttime),  // (double) create time
         (int)kp.kp_proc.p_stat,                    // (int) status
         py_name                                    // (pystr) name
     );
 
-    Py_DECREF(py_name);  // safe now, even if py_name was None
-
+    Py_DECREF(py_name);
     return py_retlist;
 }
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -214,6 +214,11 @@ psutil_proc_list_fds(pid_t pid, int *num_fds) {
     int max_size = 24 * 1024 * 1024;  // 24M
     struct proc_fdinfo *fds_pointer = NULL;
 
+    if (num_fds == NULL) {
+        errno = EINVAL;
+        return NULL;
+    }
+
     errno = 0;
     ret = proc_pidinfo(pid, PROC_PIDLISTFDS, 0, NULL, 0);
     if (ret <= 0) {

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -457,7 +457,7 @@ psutil_proc_exe(PyObject *self, PyObject *args) {
             return NULL;
         }
     }
-    return PyUnicode_DecodeFSDefaultAndSize(buf);
+    return PyUnicode_DecodeFSDefault(buf);
 }
 
 

--- a/psutil/arch/osx/proc.c
+++ b/psutil/arch/osx/proc.c
@@ -508,13 +508,14 @@ psutil_proc_memory_uss(PyObject *self, PyObject *args) {
     cpu_type_t cpu_type;
     size_t private_pages = 0;
     mach_vm_size_t size = 0;
-    mach_msg_type_number_t info_count = VM_REGION_TOP_INFO_COUNT;
+    mach_msg_type_number_t info_count;
     kern_return_t kr;
     long pagesize = psutil_getpagesize();
-    mach_vm_address_t addr = MACH_VM_MIN_ADDRESS;
+    mach_vm_address_t addr;
     mach_port_t task = MACH_PORT_NULL;
     vm_region_top_info_data_t info;
     mach_port_t object_name;
+    mach_vm_address_t prev_addr;
 
     if (! PyArg_ParseTuple(args, _Py_PARSE_PID, &pid))
         return NULL;
@@ -530,7 +531,10 @@ psutil_proc_memory_uss(PyObject *self, PyObject *args) {
 
     // Roughly based on libtop_update_vm_regions in
     // http://www.opensource.apple.com/source/top/top-100.1.2/libtop.c
-    for (addr = 0; ; addr += size) {
+    for (addr = MACH_VM_MIN_ADDRESS; ; addr += size) {
+        prev_addr = addr;
+        info_count = VM_REGION_TOP_INFO_COUNT;  // reset before each call
+
         kr = mach_vm_region(
             task, &addr, &size, VM_REGION_TOP_INFO, (vm_region_info_t)&info,
             &info_count, &object_name);
@@ -541,8 +545,15 @@ psutil_proc_memory_uss(PyObject *self, PyObject *args) {
         else if (kr != KERN_SUCCESS) {
             PyErr_Format(
                 PyExc_RuntimeError,
-                "mach_vm_region(VM_REGION_TOP_INFO) syscall failed");
+                "mach_vm_region(VM_REGION_TOP_INFO) syscall failed"
+            );
+            mach_port_deallocate(mach_task_self(), task);
             return NULL;
+        }
+
+        if (size == 0 || addr < prev_addr) {
+            psutil_debug("prevent infinite loop");
+            break;
         }
 
         if (psutil_in_shared_region(addr, cpu_type) &&

--- a/psutil/arch/posix/init.h
+++ b/psutil/arch/posix/init.h
@@ -19,7 +19,7 @@ extern PyObject *ZombieProcessError;
 
     int psutil_sysctl(int *mib, u_int miblen, void *buf, size_t buflen);
     int psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen);
-    int psutil_sysctl_argmax();
+    size_t psutil_sysctl_argmax();
 
     #if !defined(PSUTIL_OPENBSD)
 

--- a/psutil/arch/posix/net.c
+++ b/psutil/arch/posix/net.c
@@ -246,7 +246,7 @@ error:
 }
 
 static int
-append_flag(PyObject *py_retlist, const char * flag_name)
+append_flag(PyObject *py_retlist, const char *flag_name)
 {
     PyObject *py_str = NULL;
 

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -108,15 +108,20 @@ psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen) {
 }
 
 
-// Get the maximum process arguments size.
+// Get the maximum process arguments size. Return 0 on error.
 size_t
 psutil_sysctl_argmax() {
     int argmax;
     int mib[2] = {CTL_KERN, KERN_ARGMAX};
 
     if (psutil_sysctl(mib, 2, &argmax, sizeof(argmax)) != 0) {
-        PyErr_Clear();
-        psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl(KERN_ARGMAX)");
+        return 0;
+    }
+
+    if (argmax <= 0) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "sysctl(KERN_ARGMAX) return <= 0"
+        );
         return 0;
     }
 

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -62,6 +62,10 @@ psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen) {
         return -1;
     }
 
+    if (needed == 0) {
+        psutil_debug("psutil_sysctl_malloc() size = 0");
+    }
+
     while (max_retries-- > 0) {
         // zero-initialize buffer to prevent uninitialized bytes
         buffer = calloc(1, needed);
@@ -177,12 +181,23 @@ psutil_sysctlbyname_malloc(const char *name, char **buf, size_t *buflen) {
     char *buffer = NULL;
     char errbuf[256];
 
+    if (!name || !buf || !buflen) {
+        PyErr_SetString(
+            PyExc_ValueError, "psutil_sysctlbyname_malloc() invalid args"
+        );
+        return -1;
+    }
+
     // First query to determine required size.
     ret = sysctlbyname(name, NULL, &needed, NULL, 0);
     if (ret == -1) {
         snprintf(errbuf, sizeof(errbuf), "sysctlbyname('%s') malloc 1/3", name);
         psutil_PyErr_SetFromOSErrnoWithSyscall(errbuf);
         return -1;
+    }
+
+    if (needed == 0) {
+        psutil_debug("psutil_sysctlbyname_malloc() size = 0");
     }
 
     while (max_retries-- > 0) {

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -136,6 +136,13 @@ psutil_sysctlbyname(const char *name, void *buf, size_t buflen) {
     size_t len = buflen;
     char errbuf[256];
 
+    if (!name || !buf || buflen == 0) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_sysctlbyname() invalid args"
+        );
+        return -1;
+    }
+
     if (sysctlbyname(name, buf, &len, NULL, 0) == -1) {
         snprintf(errbuf, sizeof(errbuf), "sysctlbyname('%s')", name);
         psutil_PyErr_SetFromOSErrnoWithSyscall(errbuf);

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -96,7 +96,7 @@ psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen) {
 
 
 // Get the maximum process arguments size.
-int
+size_t
 psutil_sysctl_argmax() {
     int argmax;
     int mib[2] = {CTL_KERN, KERN_ARGMAX};
@@ -107,7 +107,7 @@ psutil_sysctl_argmax() {
         return 0;
     }
 
-    return argmax;
+    return (size_t)argmax;
 }
 
 

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -20,6 +20,11 @@ int
 psutil_sysctl(int *mib, u_int miblen, void *buf, size_t buflen) {
     size_t len = buflen;
 
+    if (!mib || miblen == 0 || !buf || buflen == 0) {
+        PyErr_SetString(PyExc_RuntimeError, "psutil_sysctl() invalid args");
+        return -1;
+    }
+
     if (sysctl(mib, miblen, buf, &len, NULL, 0) == -1) {
         psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl()");
         return -1;
@@ -32,6 +37,7 @@ psutil_sysctl(int *mib, u_int miblen, void *buf, size_t buflen) {
 
     return 0;
 }
+
 
 // Allocate buffer for sysctl with retry on ENOMEM or buffer size mismatch.
 // The caller is responsible for freeing the memory.

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -20,10 +20,8 @@ int
 psutil_sysctl(int *mib, u_int miblen, void *buf, size_t buflen) {
     size_t len = buflen;
 
-    if (!mib || miblen == 0 || !buf || buflen == 0) {
-        PyErr_SetString(PyExc_RuntimeError, "psutil_sysctl() invalid args");
-        return -1;
-    }
+    if (!mib || miblen == 0 || !buf || buflen == 0)
+        return psutil_badargs("psutil_sysctl");
 
     if (sysctl(mib, miblen, buf, &len, NULL, 0) == -1) {
         psutil_PyErr_SetFromOSErrnoWithSyscall("sysctl()");
@@ -48,12 +46,8 @@ psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen) {
     int ret;
     int max_retries = MAX_RETRIES;
 
-    if (!mib || miblen == 0 || !buf || !buflen) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_sysctl_malloc() invalid args"
-        );
-        return -1;
-    }
+    if (!mib || miblen == 0 || !buf || !buflen)
+        return psutil_badargs("psutil_sysctl_malloc");
 
     // First query to determine required size
     ret = sysctl(mib, miblen, NULL, &needed, NULL, 0);
@@ -140,12 +134,8 @@ psutil_sysctlbyname(const char *name, void *buf, size_t buflen) {
     size_t len = buflen;
     char errbuf[256];
 
-    if (!name || !buf || buflen == 0) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_sysctlbyname() invalid args"
-        );
-        return -1;
-    }
+    if (!name || !buf || buflen == 0)
+        return psutil_badargs("psutil_sysctlbyname");
 
     if (sysctlbyname(name, buf, &len, NULL, 0) == -1) {
         snprintf(errbuf, sizeof(errbuf), "sysctlbyname('%s')", name);
@@ -181,12 +171,8 @@ psutil_sysctlbyname_malloc(const char *name, char **buf, size_t *buflen) {
     char *buffer = NULL;
     char errbuf[256];
 
-    if (!name || !buf || !buflen) {
-        PyErr_SetString(
-            PyExc_RuntimeError, "psutil_sysctlbyname_malloc() invalid args"
-        );
-        return -1;
-    }
+    if (!name || !buf || !buflen)
+        return psutil_badargs("psutil_sysctlbyname_malloc");
 
     // First query to determine required size.
     ret = sysctlbyname(name, NULL, &needed, NULL, 0);

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -48,6 +48,13 @@ psutil_sysctl_malloc(int *mib, u_int miblen, char **buf, size_t *buflen) {
     int ret;
     int max_retries = MAX_RETRIES;
 
+    if (!mib || miblen == 0 || !buf || !buflen) {
+        PyErr_SetString(
+            PyExc_RuntimeError, "psutil_sysctl_malloc() invalid args"
+        );
+        return -1;
+    }
+
     // First query to determine required size
     ret = sysctl(mib, miblen, NULL, &needed, NULL, 0);
     if (ret == -1) {

--- a/psutil/arch/posix/sysctl.c
+++ b/psutil/arch/posix/sysctl.c
@@ -183,7 +183,7 @@ psutil_sysctlbyname_malloc(const char *name, char **buf, size_t *buflen) {
 
     if (!name || !buf || !buflen) {
         PyErr_SetString(
-            PyExc_ValueError, "psutil_sysctlbyname_malloc() invalid args"
+            PyExc_RuntimeError, "psutil_sysctlbyname_malloc() invalid args"
         );
         return -1;
     }

--- a/psutil/arch/posix/users.c
+++ b/psutil/arch/posix/users.c
@@ -31,6 +31,7 @@ teardown() {
 PyObject *
 psutil_users(PyObject *self, PyObject *args) {
     struct utmpx *ut;
+    size_t host_len;
     PyObject *py_username = NULL;
     PyObject *py_tty = NULL;
     PyObject *py_hostname = NULL;
@@ -55,11 +56,22 @@ psutil_users(PyObject *self, PyObject *args) {
         if (! py_tty)
             goto error;
 
-        if (strcmp(ut->ut_host, ":0") == 0 || strcmp(ut->ut_host, ":0.0") == 0)
+        host_len = strnlen(ut->ut_host, sizeof(ut->ut_host));
+        if (host_len == 0 ||
+            (strcmp(ut->ut_host, ":0") == 0) ||
+            (strcmp(ut->ut_host, ":0.0") == 0))
+        {
             py_hostname = PyUnicode_DecodeFSDefault("localhost");
-        else
-            py_hostname = PyUnicode_DecodeFSDefault(ut->ut_host);
-        if (! py_hostname)
+        }
+        else {
+            // ut_host might not be null-terminated if the hostname is
+            // very long, so we do it.
+            char hostbuf[sizeof(ut->ut_host)];
+            memcpy(hostbuf, ut->ut_host, host_len);
+            hostbuf[host_len] = '\0';
+            py_hostname = PyUnicode_DecodeFSDefault(hostbuf);
+        }
+        if (!py_hostname)
             goto error;
 
         py_tuple = Py_BuildValue(

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -230,7 +230,7 @@ test-command = "make -C {project} PYTHON=python PSUTIL_ROOT_DIR=\"{project}\" ci
 # Run tests on Python 3.13. On all other Python versions do a lightweight
 # import test.
 [[tool.cibuildwheel.overrides]]
-select = "cp3{8,9,10,11,12,13t,14}-* cp313-macosx"
+select = "cp3{8,9,10,11,12,13t,14}-* cp313-macosx*"
 test-extras = []
 test-command = "python -c \"import psutil; print(psutil.__version__)\""
 
@@ -238,11 +238,6 @@ test-command = "python -c \"import psutil; print(psutil.__version__)\""
 select = "cp38-macosx*"  # In macOS use Python 3.8: higher Python version hang for some reason.
 test-extras = ["test"]
 test-command = "make -C {project} PYTHON=python PSUTIL_ROOT_DIR=\"{project}\" ci-test-cibuildwheel"
-
-[[tool.cibuildwheel.overrides]]
-select = "cp3*macosx*arm64"  # XXX: skip due to segfault: https://github.com/giampaolo/psutil/issues/2661
-test-extras = ["test"]
-test-command = "python -c \"import psutil; print(psutil.__version__)\""
 
 [tool.pytest.ini_options]
 addopts = '''

